ported Razor Far to new Project System design

heejaechang opened the pull request:
for now, PR is on jason's private branch. since it is not on roslyn master. not sure how to create PR based on jason's private branch on roslyn's repro with changes in my private repo.

this is port of https://github.com/dotnet/roslyn/pull/26939 on https://github.com/dotnet/roslyn/pull/26931

heejaechang committed:
added a way to inject dynamic file to workspace for component like razor.

through IDynamicFileInfoProvider and etc, external component such as razor should be able to provide us dynamic file (such as generated c# content for cshtml file) on the fly and let us know update on the content.

renamed Column to Character since Column should be only used when it accounts "tab". and count "tab" as tab size rather than just 1 char.

heejaechang commented:
tagging @dotnet/roslyn-ide in case they want to take a look. this is not final form. working with @jasonmalinowski

heejaechang commented:
@jasonmalinowski I ported the Razor Far on your new design. 

I removed file added with SourceTextContainer to not opened by default. but made one to explicitly call Open to make it open. adding that state to existing code was a bit tricky due to batching. and state was saved as Bidirectionary <something, DocumentId>. 

so, now adding file through soucetextcontainer is same as adding file through path. one has to open a file explicitly if they want to.

but I still not sure about the batching though since in batching, I am not sure how the ordering works (like add file and open file, if add file became batched, will the system also remember open file is called, and replay that after batching end?)

anyway, my intention was making it same as regular one, but if there is something I missed let me know.

heejaechang commented on src/Features/Core/Portable/DocumentSpanExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static class DocumentSpanExtensions
```
this shows up as new file but, this is actually original file. and part of this file has moved down to Workspace layer.

heejaechang commented:
tagging @rynowak I will provide you a private soon so that you can move to this new one and start testing.

tagging @jinujoseph

heejaechang commented:
@jasonmalinowski I think current implementation gives primary buffer not secondary buffer for cshtml file in the Roslyn solution. is venus supposed to work?

heejaechang commented:
@jasonmalinowski this is legacy web project.

added through this

```
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioProject.AddSourceFile(string fullPath, Microsoft.CodeAnalysis.SourceCodeKind sourceCodeKind, System.Collections.Immutable.ImmutableArray<string> folders, Microsoft.CodeAnalysis.Host.IDocumentServiceProvider documentServiceProvider) Line 395	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.Legacy.AbstractLegacyProject.AddFile(string filename, Microsoft.CodeAnalysis.SourceCodeKind sourceCodeKind) Line 127	C#
 	Microsoft.VisualStudio.LanguageServices.CSharp.dll!Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim.CSharpProjectShim.OnSourceFileAdded(string filename) Line 51	C#
```

![image](https://user-images.githubusercontent.com/1333179/45346822-34f82b80-b55f-11e8-9ddf-f1a9d6148447.png)

rather than file is being added through ContainedLanguage?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -382,16 +381,6 @@ private void OnBatchScopeDisposed()
 
-                foreach (var (documentId, textContainer) in documentsToOpen)
-                {
-                    _workspace.ApplyChangeToWorkspace(w => w.OnDocumentOpened(documentId, textContainer));
```
As discussed earlier: we do still want this for the classic contained language case, so best to just add a flag I think.

heejaechang replied:
sure. I will add a dictionary that tracks the open state.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -1012,8 +1006,9 @@ public void RemoveTextContainer(SourceTextContainer textContainer)
-                                w.OnDocumentClosed(documentId, new SourceTextLoader(textContainer, filePath: null));
+                            _project._workspace.ApplyChangeToWorkspace(w =>
+                            {
+                                // removing document automatically close the file if it is opened
```
Does it? I see OnDocumentRemoved calling CheckDocumentCanBeRemoved which is checking open/closed state.

heejaechang replied:
interesting. what I looked was this (ClearDocumentData) which calls
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Workspace_Editor.cs,106 

that code makes sure it removes document from OpenFile tracking. which is only called from RemoveDoc/AdditionalDoc..

anyway, I will put it back

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs:
```diff
@@ -0,0 +1,85 @@
+                public async Task<ImmutableArray<DocumentSpan>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken)
+                {
+                    var containedDocument = ContainedDocument.TryGetContainedDocument(document.Id);
+                    if (containedDocument == null)
```
How could this ever happen? Seems like we'd want to throw, since something would be most amiss?

heejaechang replied:
sure. can make it to throw

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Venus/ContainedLanguage.cs:
```diff
@@ -133,7 +132,13 @@ internal partial class ContainedLanguage<TPackage, TLanguageService>
+                documentServiceProvider: ContainedDocument.DocumentServiceProvider.Instace);
+
+            // all contained document is opened by default
+            Workspace.OpenDocument(documentId);
```
This isn't technically safe, since if AddSourceTextContainer was batched, it's still not here yet. I think we're best off just adding the flag and shuttling around the value somewhere.

heejaechang replied:
sure.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs:
```diff
@@ -0,0 +1,85 @@
+                    var builder = ArrayBuilder<DocumentSpan>.GetInstance();
+                    foreach (var span in spans)
+                    {
+                        foreach (var primarySpan in editorSnapshot.MapToSourceSnapshots(span.ToSpan()))
```
Is this mapping in the correct direction? I would have presumed this was taking a location in the secondary buffer and mapped to the primary, and technically editorSnapshot doesn't have to be an editor snapshot at all. This might be relying on the fact that the data buffer/document buffer have the same contents but it's kinda strange...

heejaechang replied:
I believe it is on right direction.

..

this is old code we use (which I also verified it works)
TryMapSpanFromSecondaryBufferToPrimaryBuffer (http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Extensions/VsTextSpanExtensions.cs,12)

I basically ported editor implementation
http://index/?query=MapSecondaryToPrimarySpan&rightProject=Microsoft.VisualStudio.Editor.Implementation&file=VsTextBufferCoordinatorAdapter.cs&line=177

but, I might wrongly cast it to IProjectionBuffer. will take a look.

heejaechang replied:
I think name of editorSnapshot confusing it. I used editorSnapshot as a way to say it is not Roslyn Buffer. but I can rename it to secondaryBufferSnapshot.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs:
```diff
@@ -0,0 +1,85 @@
+                            // and consider span that is not split
+                            if (primarySpan.Length != span.Length)
+                            {
+                                continue;
```
This would break the contract that MapSpansAsync would return a 1:1 for the original span going in and the resulting mapped span. In other words, if you put four spans in and get three out, you won't even know which of the four failed.

heejaechang replied:
sure. I can keep the 1:1 mapping. either Document being null or null entry.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs:
```diff
@@ -0,0 +1,85 @@
+                                continue;
+                            }
+
+                            builder.Add(new DocumentSpan(document, primarySpan.Span.ToTextSpan()));
```
This doesn't seem right: the spans here now don't correspond to the document itself. The document in this case is the secondary buffer. More generally, it seems like this should be returning some pair more like "filename + span", because in the closed file case we're trying to fix, the .cshtml will have no traditional document representation.

jasonmalinowski replied:
Think about this this way: DocumentSpan (for perf reasons) doesn't have an assert that the span is within the document's text. If it did...this would be throwing.

heejaechang replied:
with Ryan's change, all razor file will have document. and won't have filepath. currently it works since filepath of the document we have is for the cshtml file. that's why it currently works.

can't do other thing since this whole our system (find all references) expects document unfortunately.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Host/DocumentService/ISpanMapperService.cs:
```diff
@@ -0,0 +1,31 @@
+        /// <param name="spans">Spans in the document</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Return mapped span. order of result should be same as the given span</returns>
+        Task<ImmutableArray<DocumentSpan>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken);
```
Commented on the implementation, but it doesn't seem that DocumentSpan is the right currency here. In the case we're trying to address, there's a .cshtml file on disk, and we have a file being added for that source. But we _don't_ actually have a document _for the .cshtml file itself_.

heejaechang replied:
the work Ryan is doing is adding Document for all closed files.

rynowak replied:
> But we don't actually have a document for the .cshtml file itself.

Right, I agree with @jasonmalinowski - you guys will have a `Document` that contains the generated code (C#), but not a document that contains the original source code (.cshtml)

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs:
```diff
@@ -71,22 +78,39 @@ internal DocumentInfo(DocumentAttributes attributes, TextLoader loader)
             DocumentAttributes attributes = null,
-            Optional<TextLoader> loader = default)
+            Optional<TextLoader> loader = default,
+            IDocumentServiceProvider documentServiceProvider = null)
```
Should be Optional<IDocumentServiceLoader> as you'd have no way to With(documentServiceProvider: null).

heejaechang replied:
sure.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs:
```diff
@@ -40,44 +39,52 @@ internal partial class TextDocumentState
+        /// <summary>
+        /// A <see cref="IDocumentServiceProvider"/> associated with this document
+        /// 
+        /// not every <see cref="Workspace"/> or <see cref="Document"/> supports 
```
Should we just sub in a no-op provider that always returns null?

heejaechang replied:
yep. I agree.

jasonmalinowski replied:
Comments need updating then. (You did fix the code.)

heejaechang replied:
ah. yes.  I will change the comment.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/FindReferences/Entries/AbstractDocumentSpanEntry.cs:
```diff
@@ -86,6 +104,25 @@ public override bool TryCreateColumnContent(string columnName, out FrameworkElem
+
+            public static async Task<DocumentSpan> MapAndGetFirstAsync(DocumentSpan documentSpan, CancellationToken cancellationToken)
+            {
+                var service = documentSpan.Document.State.DocumentServices?.GetService<ISpanMapperService>();
```
We should make an internal method on Document so we're not grovelling throguh Document.State.DocumentServices like this.

jasonmalinowski replied:
(also consider maybe just making that a place that does the null check for DocumentServices)

heejaechang replied:
sure. i can make it exposed on Document. I might put DefaultImplementation of DocumentServices which return null for all services so that one doesn't need to check null.

jasonmalinowski reviewed (dismissed):
Only fatal flaw I see is the ISpanMapperService doesn't seem right: spans can map to files that don't really have a Document, so DocumentSpan isn't the right thing to return. If nothing else, right now it looks like we're returning a document that happens to have the right filename but the spans wouldn't line up, and anybody who tried using that document/span combination is going to explode.

Otherwise, we should just put back the ability for SourceTextContainer to open files because there is some problems here. Just threading through the flag should be pretty simple.

heejaechang commented:
@jasonmalinowski will make that change.

rynowak commented:
Thanks @heejaechang - I'll catch up on these changes and the discussions taking place here tomorrow.

One other issue that I want to raise for discussion (this seems like the right time) - we originally debated if the right contract for a 'closed' razor-generated-code file is similar to `TextLoader` or similar to `SourceTextContainer`.  We went with something similar to `SourceTextContainer`. I don't think either approach causes issues for us, I just wanted to mention it in case it's significant in the context of this PR.

heejaechang commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -389,7 +390,13 @@ private void OnBatchScopeDisposed()
                 {
-                    _workspace.ApplyChangeToWorkspace(w => w.OnDocumentClosed(documentId, textLoader));
+                    _workspace.ApplyChangeToWorkspace(w =>
+                    {
```
I am not sure how VisualStudioProject make sure workspace doesn't mutate underneath since open/close state is mutable state on workspace.

heejaechang commented:
@rynowak now it should work with latest drop on cpvsbuild\drops\vs\ml branch

can you try this bits?

rynowak commented:
Sure, will likely be tomorrow but I'll get on it.

heejaechang commented:
@rynowak @jasonmalinowski 
> one other issue that I want to raise for discussion (this seems like the right time) - we originally debated if the right contract for a 'closed' razor-generated-code file is similar to TextLoader or similar to SourceTextContainer. We went with something similar to SourceTextContainer. I don't think either approach causes issues for us, I just wanted to mention it in case it's significant in the context of this PR.

ya, I remember that. but for now, let's keep sourcetext container way since it provides a way to know that content has changed but TextLoader doesn't. since content is not backed by a file, we can't use file tracker to monitor changes. so sourcetext is simpler way for now. 

since SourceTextContainer is just an abstraction, we could have the container that is not backed by memory so we don't hold every secondary buffer in memory. (and use like memory map file and etc)

rynowak commented:
@heejaechang  - I tried this out on the latest ML 16.0.0 Preview 1.0 [28112.4004.ml] and it's having some trouble loading projects. I haven't tried to root cause what the problem is. 

So, I'm only testing some basic stuff right now because it's having problems loading things. 

-- 

The next problem is that though I'm registering a `SourceTextContainer`... nothing subscribes to the changed event unless I specify that it's **open**. 

https://github.com/jasonmalinowski/roslyn/pull/2/files#diff-276c61ed00880328a9f6cdd152502e9bR907

If I hack around this by adding the document as **open** then FAR works largely as expected.

heejaechang commented:
I see. I will take a look at that.


  *   heejae

From: Ryan Nowak <notifications@github.com>
Sent: Friday, September 14, 2018 4:59 PM
To: jasonmalinowski/roslyn <roslyn@noreply.github.com>
Cc: HeeJae Chang <Heejae.Chang@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [jasonmalinowski/roslyn] ported Razor Far to new Project System design (#2)


@heejaechang<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fheejaechang&data=02%7C01%7CHeejae.Chang%40microsoft.com%7C56fe7747c71c44300c6508d61a9e00b9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636725663248395752&sdata=QDMAalgJ8oAENGwkKAVJjoxTZErI%2B0%2BIXyKdz11JkzU%3D&reserved=0> - I tried this out on the latest ML 16.0.0 Preview 1.0 [28112.4004.ml] and it's having some trouble loading projects. I haven't tried to root cause what the problem is.

So, I'm only testing some basic stuff right now because it's having problems loading things.

--

The next problem is that though I'm registering a SourceTextContainer... nothing subscribes to the changed event unless I specify that it's open.

https://github.com/jasonmalinowski/roslyn/pull/2/files#diff-276c61ed00880328a9f6cdd152502e9bR907<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjasonmalinowski%2Froslyn%2Fpull%2F2%2Ffiles%23diff-276c61ed00880328a9f6cdd152502e9bR907&data=02%7C01%7CHeejae.Chang%40microsoft.com%7C56fe7747c71c44300c6508d61a9e00b9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636725663248395752&sdata=jTYRU6GpYh4yJjyayzDM6%2BIYAxy89j7MjKNNcC7PY34%3D&reserved=0>

If I hack around this by adding the document as open then FAR works largely as expected.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjasonmalinowski%2Froslyn%2Fpull%2F2%23issuecomment-421513228&data=02%7C01%7CHeejae.Chang%40microsoft.com%7C56fe7747c71c44300c6508d61a9e00b9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636725663248395752&sdata=NEVeGmPcfbq06k7jVkw%2BG%2BBjlos67zKQQE%2B2SFn%2BdZE%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABRXu4UgSWl5iDp56oCzfe0FU15h5d0Kks5ubEKzgaJpZM4Wh8N2&data=02%7C01%7CHeejae.Chang%40microsoft.com%7C56fe7747c71c44300c6508d61a9e00b9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636725663248552012&sdata=WDSeLh5aH9EJvS8jDi1HmYe6DPwieKDrTo%2BtWzAC4%2FU%3D&reserved=0>.

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs:
```diff
@@ -36,6 +39,10 @@ internal interface IWorkspaceProjectContext : IDisposable
+        void AddSourceTextContainer(SourceTextContainer textContainer, string fullPath, IEnumerable<string> folderNames = null, SourceCodeKind sourceCodeKind = SourceCodeKind.Regular);
+        void RemoveSourceTextContainer(SourceTextContainer textContainer);
+        Task AddNonSourceFileAsync(string filePath, bool isInCurrentContext = true, IEnumerable<string> folderNames = null, CancellationToken cancellationToken = default);
+        Task RemoveNonSourceFileAsync(string fullPath, CancellationToken cancellationToken);
```
Looks these these are the two new APIs yes? @davkean

Some questions:

Do these need to be async? I don't remember that as something we discussed
Do we still need `AddSourceTextContainer`/`RemoveSourceTextContainer`?

heejaechang replied:
@jasonmalinowski asked me to make it Async. he said, he is planing to change all existing ones to Async. SouceTextContainers are removed.

rynowak replied:
OK, just checking. As long as @davkean and @jasonmalinowski are in agreement, then this API seems like what I expect.

davkean replied:
Please make them async all at once, not piecemeal - we call these without async context.

jasonmalinowski replied:
The challenge is this is us exposing asynchrony through that @rynowak's code is being allowed to have. Either we are going to [a] be blocking your thread on @rynowak's async call, [b] asking him to not really be async, or [c] do a three-way insertion change when we want to fix this if we don't just make it async now.

heejaechang replied:
let me know what you want on the async side.

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -448,6 +467,44 @@ public void RemoveAdditionalFile(string fullPath)
+                // we should have metadata on provider that we can filter through file extension, but for now, 
+                // just go through all providers
+                var fileInfo = await provider.Value.GetSourceFileInfoAsync(filePath, cancellationToken).ConfigureAwait(false);
+                if (filePath == null)
```
It looks like this `if` should be `fileInfo` instead of `filePath`

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -448,6 +467,44 @@ public void RemoveAdditionalFile(string fullPath)
+            foreach (var provider in _sourceFileInfoProviders)
+            {
+                var fileInfo = await provider.Value.GetSourceFileInfoAsync(filePath, cancellationToken).ConfigureAwait(false);
+                if (filePath == null)
```
Some thing, this should be `fileInfo`

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectFactory.cs:
```diff
@@ -16,12 +13,19 @@ internal sealed class VisualStudioProjectFactory
         [ImportingConstructor]
         // TODO: remove the AllowDefault = true on HostDiagnosticUpdateSource by making it a proper mock
-        public VisualStudioProjectFactory(VisualStudioWorkspaceImpl visualStudioWorkspaceImpl, [Import(AllowDefault = true)] HostDiagnosticUpdateSource hostDiagnosticUpdateSource)
+        // TODO: we might want to add metadata on ISourceFileInfoProvider to avoid unnecessary realization of service
```
We will definitely require this - our assembly needs to be loaded in just web scenarios

jasonmalinowski replied:
Indeed, this would outright block in RPS, and I think getting an exception for that would be pretty tricky.

rynowak replied:
We already get mucho hate mail from them due to this editor thing that has been blocked for a year 😆  - I'd really like to stay off of their radar.

heejaechang replied:
:) sure.

rynowak commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,58 @@
+    /// 
+    /// none of info returned from this can be dynamic. it should always return same data
+    /// </summary>
+    internal abstract class SourceFileInfo
```
This seems basically like what we discussed. 👍 Some comments

rynowak commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <summary>
+        /// unique id to identify this file among other files in a project. it should be unique in the project
+        /// </summary>
+        public abstract string Id { get; }
```
Are we supposed to generate this? I think we would end up doing like `Guid.NewGuid()` or something silly like that

heejaechang replied:
I can remove this. thought might be useful when debugging. but can be removed.

rynowak commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,58 @@
+        /// 
+        /// this can be used with something like editorconfig to find out right options for this source file.
+        /// </summary>
+        public abstract string FilePath { get; }
```
So at least for right now, we have to make sure the documents added by us in the Workspace don't actually use the filename. I assume we would provide this as null for now?

heejaechang replied:
I am not sure, I was expecting something like for

c:\aaa\bbbg.cshtml you would add something like
c:\aaa\bbbg.cshtml.generated.cs or something like that. which doesn't actually exist in file system. but there at same directory as csthml and has cs extension, so that editorconfig work?

rynowak replied:
So we could definitely provide the name of what the generated file would be during the build, I have some questions about what that would *do* - but it's not a blocker for us.

heejaechang replied:
giving it null will work as well. it won't crash.

jasonmalinowski replied:
I'd expect we don't offer a file path until the point we have a damn good reason to add it; somebody's going to use it wrong almost immediately. :smile:

heejaechang replied:
so, if we want to support analyzers and fixer stuff in future in razor file. we need to make it work with editorconfig. so, if we don't let it have a fake file path, then, we probably need to add some kind of abstraction so that it gets editorconfig from something for non-source file path. and user is expected to add cs/vb/dotnet code styles on editorconfig for cshtml/vbhtml?

rynowak commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <summary>
+        /// return TextLoader to get content of this source file
+        /// </summary>
+        public abstract TextLoader TextLoader { get; }
```
I think we're missing `IDocumentServiceFactory` here as well. 

Since `TextLoader` and `IDocumentServiceFactory` will change at the same time, we need to express that in a thread-safe way through the API.

heejaechang replied:
ah.. yep. I forgot to add the most important one :) let me add it.

rynowak commented:
@heejaechang - if I want to try this out, what VS do I need?

heejaechang commented:
@jasonmalinowski is it still only work with ml build? or d16.0stg work with this?

rynowak commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(string fullPath, CancellationToken cancellationToken);
```
I will also need some way to know what project this is associated with - project path or projectId

jasonmalinowski commented on src/EditorFeatures/Core/Shared/Extensions/SnapshotSpanExtensions.cs:
```diff
@@ -25,6 +25,12 @@ public static ITrackingSpan CreateTrackingSpan(this SnapshotSpan snapshotSpan, S
 
+        public static LinePositionSpan ToLinePositionSpan(this SnapshotSpan snapshotSpan)
+        {
+            snapshotSpan.GetLinesAndColumns(out var startLine, out var startChar, out var endLine, out var endChar);
```
Are the things coming out of this columns or character index/offsets? The method implies columns, the usage implies characters, so the moment you have a tab this is going to go very sideways.

heejaechang replied:
I am not changing any of this line/column logic from existing one. by looking at the code, everything is done by position calculation. so I dont believe Column actually convert tab to tab size. Column and char look like used interchangeably. if that is confusing, sure, I can rename GetLinesAndColumns to GetLinesAndCharacters.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(ProjectId projectId, string projectFullPath, string fileFullPath, CancellationToken cancellationToken);
```
I'd avoid a project file path: if/when you do the separate project for these, you won't have a valid path anymore. (Or, you'll have two projects with the same path and that won't be useful either.)

rynowak replied:
@jasonmalinowski - I asked for this. Since this service is MEF-imported globally, we need to know what project you're asking for. We create the `SourceFileInfo` when you ask for it, and then we have to associated it later with our state.

heejaechang replied:
let me know if you guys want to change this.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+    /// SourceFileInfo is random name I choose. we might want to chagne this to something like temporarySourceInfo,
+    /// InterimSourceInfo, GeneratedSourceInfo and etc.
+    /// 
+    /// none of info returned from property can be dynamic. it should always return same data
```
Should just this be returning a class w/ readonly members instead of using inheritance? The TextLoader and  IDocumentServiceProvider already allow for deferred providing of data.

heejaechang replied:
if we go that route, we probably want to get new SourceFileInfo per snapshot. right now, SourceFileInfo is shared between multiple document snapshots.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+        /// </summary>
+        /// <remarks>
+        /// we probably want to change this to Async and give in cancellation token. 
+        /// for now, it is not, since FileChange event is synchronous and doesn't provide cancellationToken
```
Add it now or never.

heejaechang replied:
@rynowak do you want it to be async or synch is fine?

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+
+        /// <summary>
+        /// indicates that content of this source file has updated. new content will be retrieved by
+        /// TextLoader.LoadTextAndVersionAsync
```
Is the contract that a new TextLoader is provided, or the previous TextLoader will return different values?

heejaechang replied:
it is supposed to return new textloader and document service provider that is for a specific document snapshot. 

one can return same textloader as long as it return new content that matches documentServiceProvider as a implementation detail.

but the engine will call textLoader.GetContent to get new content.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs:
```diff
@@ -390,5 +400,17 @@ public virtual async Task<VersionStamp> GetTopLevelChangeTextVersionAsync(Cancel
+
+            public TService GetService<TService>() where TService : IDocumentService
+            {
+                return default;
```
Are we not constraining this to returning reference types? Feels like that might help down the road lest we end up with strange casting things.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs:
```diff
@@ -40,44 +39,52 @@ internal partial class TextDocumentState
             this.textAndVersionSource = textAndVersionSource;
 
+            Attributes = attributes;
+            DocumentServices = documentServiceProvider ?? NoOpDocumentServiceProvider.Instance;
```
nit: it'd be nice to sort these the same way as constructor arguments.

heejaechang replied:
sure.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/TextDocument.cs:
```diff
@@ -51,6 +52,14 @@ internal TextDocument(Project project, TextDocumentState state)
+        /// <summary>
+        /// A <see cref="IDocumentServiceProvider"/> associated with this document
+        /// 
+        /// not every <see cref="Workspace"/> or <see cref="Document"/> supports 
```
Comment needs updating now.

heejaechang replied:
not sure what part needs to be updated?

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISourceFileInfoProvider
```
What is the plan for lazy loading these with MEF metadata?

heejaechang replied:
probably we will add metadata for file extensions, and only ones that claims to support matching file extension will be realized?

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(ProjectId projectId, string projectFullPath, string fileFullPath, CancellationToken cancellationToken);
```
What happens if we need two files for some reason?

heejaechang replied:
1 non source file returning 2 source files? since.. file is not actually meaningful, I guess they can return a file that contains multiple types in it?

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+        /// the event arg here (FilePath) is the original non source file full path given to
+        /// GetSourceFileInfoAsync
+        /// </summary>
+        public event EventHandler<string> Updated;
```
Does this give a way to remove? Or raise this and have GetTextLoaderAndDocuments return null?

rynowak replied:
**real remove** is done through the workspace project context - my plan was to set the source text to 'empty' and raise update when the document opens in the editor.

heejaechang replied:
yep. that was my understanding from the meeting as well.

heejaechang replied:
until we unify open/close document.

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Host/DocumentService/IDocumentServiceProvider.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        /// <summary>
+        /// Gets a document specific service provided by the host identified by the service type. 
+        /// If the host does not provide the service, this method returns null.
```
Commented elsewhere too, but if you want null we should add a where `TService : class` constraint.

rynowak replied:
+1 this is just generally a good idea, I don't think we have much of a use case for structs as document services.

heejaechang replied:
sure

jasonmalinowski commented on src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs:
```diff
@@ -158,6 +160,22 @@ public void AddAdditionalFile(string filePath, bool isInCurrentContext = true)
+
+        public Task AddNonSourceFileAsync(string filePath, bool isInCurrentContext = true, IEnumerable<string> folderNames = null, CancellationToken cancellationToken = default)
+        {
+            // isInCurrentContext might not needed? AddSourceFile doesn't use it.
```
It never should have been added, so you mind as well remove it now. I can't remove it for back-compat concerns.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs:
```diff
@@ -91,6 +93,12 @@ public interface IContext : IDisposable
 
+            /// <summary>
+            /// Starts watching content changes through SourceFileInfo abstraction rather than going directly through a physical file. 
+            /// this lets us deal with virtual file that doesn't actually exist in the file system
```
This is really intended to be abstract from the workspace stuff and is used in other places. If we want to reuse the abstraction let's discuss.

heejaechang replied:
sure. I wanted to share remove file implementation. I can make it not share that implementation and do content change tracking differently.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs:
```diff
@@ -198,6 +209,26 @@ public IFileWatchingToken EnqueueWatchingFile(string filePath)
                 return token;
             }
 
+            public IFileWatchingToken EnqueueWatchingFile(SourceFileInfo fileInfo)
```
Commented above, but since this method doesn't actually need any thing from this, we can keep this impl/trick in VisualStudioProject.

(longer term, other parts of VS are moving over to consume this newer file watching abstraction, so we want to keep it pure)

heejaechang replied:
sure.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs:
```diff
@@ -189,6 +198,8 @@ public IFileWatchingToken EnqueueWatchingFile(string filePath)
                     ErrorHandler.ThrowOnFailure(service.AdviseFileChange(filePath, (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time), this, out cookie));
 
+                    // not sure why this lock is here.
+                    // might be left out from other usage?
```
Ah, yes, please break the removal into a separate PR against my branch.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs:
```diff
@@ -162,9 +170,10 @@ public void Dispose()
                             ErrorHandler.ThrowOnFailure(service.UnadviseDirChange(_directoryWatchCookie));
                         }
 
+                        // shouldn't _activeFileWatchingTokens need a lock or need local copy before the queue?
```
Only if you're watching/unwatching files after you called Dispose(). Which is a good call: those methods should throw in that case because don't do that.

heejaechang replied:
my worry was more on data corruption where other thread modified the HashSet but since Dispose doesn't have memory barrier, not all that changes are synched to the thread Disposed is called?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -408,9 +425,11 @@ public void AddSourceFile(string fullPath, SourceCodeKind sourceCodeKind = Sourc
 
-        public DocumentId AddSourceTextContainer(SourceTextContainer textContainer, string fullPath, SourceCodeKind sourceCodeKind = SourceCodeKind.Regular, ImmutableArray<string> folders = default)
+        public DocumentId AddSourceTextContainer(
+            SourceTextContainer textContainer, string fullPath,
```
If we're going to break into multiple lines, let's do a single parameter per line.

heejaechang replied:
sure

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -808,6 +851,11 @@ private sealed class BatchingDocumentCollection
+            /// <summary>
+            /// the map of non source file paths to <see cref="SourceFileInfo"/> that provide actual generated content
+            /// </summary>
+            private readonly Dictionary<string, SourceFileInfo> _nonSourceFilePathToSourceFileInfo = new Dictionary<string, SourceFileInfo>(StringComparer.OrdinalIgnoreCase);
```
This shouldn't be here: this class exists precisely to avoid duplication for regular source files and additional files. Observe the two instances in the parent. I think you might want a third instance, and the extra tracking just being held by VisualStudioProject?

heejaechang replied:
sure.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -933,6 +985,58 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
+                {
+                    if (_documentPathsToDocumentIds.ContainsKey(filePath))
+                    {
+                        throw new ArgumentException($"'{filePath}' has already been added to this project.", nameof(filePath));
```
So this throw here is particularly funky: I realize it's bad, but it means if @rynowak screws up, CPS is getting an exception thrown to them. We'll need to revisit this.

rynowak replied:
This isn't us though, we don't call this API. `filePath` here is coming from the project system

heejaechang replied:
this being throw on the file path is not different than AddSourceFile though? all input should comes from same one as AddSourceFile so, either both shouldn't do it or both should be fine?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -933,6 +985,58 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
+
+                    _documentPathsToDocumentIds.Add(filePath, documentId);
+
+                    _project._fileWatchingTokens.Add(documentId, _project._documentFileChangeContext.EnqueueWatchingFile(fileInfo));
```
Commented elsewhere, if you want to use the IFileWatchingToken interface as an abstraction, I think it makes more sense just to have a separate implementation which we create here.

heejaechang replied:
sure.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -1126,6 +1245,33 @@ public void ProcessFileChange(string fullFilePath)
+            {
+                // if we want "editorconfig" to work for this file, we might at least want something here? even if
+                // the file doesn't actually exist in file system?
+                var filePath = fileInfo.FilePath ?? string.Empty;
```
Not sure why we'd sub *in* string.Empty. Normally we use real nulls...

heejaechang replied:
sure.

jasonmalinowski reviewed (commented):
Leaving a bunch of comments. High level this does match what I recall from our conversation. I would like to see if we can clean up the implementation bits in VisualStudioProject, but none of that should block progress or other cleanup.

rynowak commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis
```
Just thought of while implementing - should this API notify us when a `SourceFileInfo` is removed? (or make it disposable).

heejaechang replied:
we can do it if that is necessary but not having Dispose is usually make things easier. do you have a reason why want to get noticed when file is removed? IDocumentServiceProvide being specific to 1 snapshot (Document), not enough for state management?

heejaechang replied:
I think I shouldn't say 1 snapshot. I meant 1 green node (DocumentState).

rynowak commented:
So my current blocking issue:
```
   at Roslyn.Utilities.Contract.Fail(String message)
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition, String message)
   at Microsoft.CodeAnalysis.Remote.SolutionCreator.<UpdateDocumentInfoAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task **task)**
```

SolutionCreator@L78 throws when we update our documents.

rynowak commented:
Another blocking issue - this code will throw when a Razor file gets opened in the editor. 

This change adds the 'filePath' to the `_documentPathsToDocumentIds` here https://github.com/jasonmalinowski/roslyn/pull/2/files#diff-276c61ed00880328a9f6cdd152502e9bR995

We need it to not conflict with the contained language buffer. 

----

**Edit:** So this issue seems somewhat blocking (as far as me consuming the prototype) then since the change is also processed based on the file name.

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -933,6 +985,58 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
+            {
+                lock (_project._gate)
+                {
+                    if (!_nonSourceFilePathToSourceFileInfo.TryGetValue(filePath, out var fileInfo))
```
I don't think there's any code that adds to `_nonSourceFilePathToSourceFileInfo` so the code that processes changes will never see a change.

heejaechang replied:
:) I just realized as well :) sorry about that.

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -1064,21 +1174,30 @@ public void ProcessFileChange(string fullFilePath)
+                            }
+                            else
+                            {
+                                documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, document.Folders);
```
There's another blocking issue here. This document is already part of the workspace and we are processing a change. This method will return a `DocumentInfo` with a new `DocumentId` but we need to use `document.Id` (see line 1190). This will cause line 1202 to throw because the solution thinks it's not part of the solution.

heejaechang replied:
yep. my bad.

heejaechang commented:
@rynowak 
can you give me more info on this?

 at Roslyn.Utilities.Contract.Fail(String message)
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition, String message)
   at Microsoft.CodeAnalysis.Remote.SolutionCreator.<UpdateDocumentInfoAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task **task)**

I am not sure what has failed.

heejaechang commented:
@rynowak  can you try again? I will address other feedbacks meanwhile, I checked in changes to at least unblock you except the Contrail.Fail above. 

I am not sure how that can happen. can you give me a dump or more info?

rynowak commented:
@heejaechang - the error is coming from here: https://github.com/jasonmalinowski/roslyn/blob/free-threaded-project-system-api/src/Workspaces/Remote/Core/Services/SolutionCreator.cs#L78 

If this still reproduces when I update to your latest, I will get a dump or come by and show you so we can look at it together.

rynowak commented on src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs:
```diff
@@ -0,0 +1,61 @@
+        /// and not give in what old and new text/version were when event happened but just FilePath.
+        /// 
+        /// the event arg here (FilePath) is the original non source file full path given to
+        /// GetSourceFileInfoAsync
```
This event also has to be abstract. If my code doesn't define it, I can't raise it.

heejaechang replied:
added OnUpdate method so that you can raise it.

rynowak replied:
That works too 👍

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -933,6 +996,65 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
 
+            public Task AddSourceFileInfoAsync(string filePath, SourceFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.IsDefault ? null : (IEnumerable<string>)folders);
```
This call now throws is `fileInfo.FilePath` is null. I'm working around this for now in my code by providing the file path.

heejaechang committed:
added APIs for project system

heejaechang committed:
resolve 2 issues we found while prototyping

jasonmalinowski committed:
Populate VisualStudioWorkspace's solution file path

We stopped populating this, which breaks various things like Live Unit
Testing and some of our persistence stuff.

davkean commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -498,20 +507,27 @@ public void AddDynamicFile(string filePath, ImmutableArray<string> folders)
                 // where filePath points to dynamic file such as cshtml and etc
                 // for now, we use GetAwaiter().GetResult() until this method itself becomes Async
-                var fileInfo = provider.Value.GetDynamicFileInfoAsync(projectId: Id, projectFilePath: _filePath, filePath: filePath, CancellationToken.None).GetAwaiter().GetResult();
+                var fileInfo = _threadingContext.JoinableTaskFactory.Run(() => provider.Value.GetDynamicFileInfoAsync(projectId: Id, projectFilePath: _filePath, filePath: filePath, CancellationToken.None));
```
Ugh. Didn't realize that's what you are doing under the covers, I thought you were going to implement a free-threaded implementation. The second we consume this on a background - this is going to cause ThreadPool exhaustion. 

Do we have a bug tracking adding async everywhere? We'll need that before I call this on a background thread.

davkean replied:
@jasonmalinowski ^

heejaechang replied:
@jasonmalinowski ?  

also, added trakcing issue
https://github.com/dotnet/roslyn/issues/30533

rynowak replied:
Jason want the APIs they provided to us to be async since that's the direction these things are moving in - I don't need it and my implementation is free threaded.

heejaechang replied:
@davkean @rynowak 
> a free-threaded implementation

this is free thread implementation, I am not sure what you meant by it not free-thread.

I am using JTF.Run here to call async method in a blocking way without blocking UI thread if it ever used on UI thread.

Wait or Result will cause dead lock if this method is called from UI thread, and the async method ever try to switch to UI thread.

let me tag @AArnott to confirm.

davkean replied:
If this code _ever_ calls code that switches to the UI thread - I cannot call this off the UI thread. I will run out of threads very quickly.

AArnott replied:
@heejaechang Using JTF.Run instead of Task.Wait or Task.Result is indeed a good way to go. But async methods are generally allowed to have thread affinity, so the async work you do inside *may* need the UI thread (now or in the future, directly or transitively). So any form of blocking a thread while async works executes should be avoided if possible to avoid threadpool exhaustion when called from a threadpool thread as @davkean said, and to avoid UI delays if called from the UI thread. And if this code may execute multiple times in parallel, that makes the threadpool exhaustion risk much greater. 
Remember that netbooks may have only *two cores*, so .NET will only allocate 2 thread pool threads on those machines before adding 1 second delays before reluctantly adding each new threadpool thread. So [threadpool exhaustion](https://github.com/Microsoft/vs-threading/blob/master/doc/threadpool_starvation.md) is a very real concern that requires constant vigilance.

AArnott replied:
As this method is internal (the class is internal and implements no interfaces), the [VSTHRD102](https://github.com/Microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD102.md) analyzer should emit a warning here advising you to avoid JTF.Run since you could expose this asynchronously. What prevents you from doing this?

heejaechang replied:
long story. changing the existing call paths between 2 teams require some coordination. so it is an intermediate thing which we will remove once we move existing call paths to async, until then, doing this pattern.

right now, callee (implementation) of the async method in JTF.Run do work in a synchronous way, so JTF.Run should be fine here.

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -1058,27 +1246,44 @@ public void ProcessFileChange(string fullFilePath)
+                            }
+                            else
+                            {
+                                var fileInfo = _project._threadingContext.JoinableTaskFactory.Run(() => fileInfoProvider.GetDynamicFileInfoAsync(_project.Id, _project._filePath, fullFilePath, CancellationToken.None));
```
@heejaechang  - this isn't right. 

This is passing in the mangled file path. Ex:
C:\Users\rynowak\Source\Repos\WebApplication6\WebApplication6\Views\Shared\Error.cshtml#Generated#
instead of 
C:\Users\rynowak\Source\Repos\WebApplication6\WebApplication6\Views\Shared\Error.cshtml

heejaechang replied:
oops. my bad. I will make it to give you right path.

heejaechang replied:
fixed.

rynowak commented:
@heejaechang - I've updated on top of this and it's a little simpler to implement from my POV. There's one blocking issue I've found here https://github.com/jasonmalinowski/roslyn/pull/2/files#r225397963

If I work around that in my code, I end up back where we were a few days ago. For some reason the initial push of generated code is ignored, and then after that things work fine. I'll come by tomorrow so we can review the use of `FilePath` again, I suspect that's where I went wrong. 

In case you're curious what our side looks like: https://github.com/aspnet/Razor/compare/rynowak/workspace-integration?expand=1

rynowak commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -1058,27 +1246,44 @@ public void ProcessFileChange(string fullFilePath)
+                            {
+                                var fileInfo = _project._threadingContext.JoinableTaskFactory.Run(() => fileInfoProvider.GetDynamicFileInfoAsync(_project.Id, _project._filePath, fullFilePath, CancellationToken.None));
+
+                                textLoader = fileInfo.TextLoader;
```
To add another datapoint here about `JoinableTaskFactory.Run` - this means that each time we update a document, it has to transition to the fg thread and then back. When testing this on a larger solution this is adding some noticable significant overhead;

heejaechang replied:
I just looked JTF.Run source and i dont see anything about switching to UI thread?

just to make sure let me ask @AArnott 

Andrew, if I use JTF.Run(async () => await ...), I get that it will block current thread, but will it switch to UI thread?

I thought JTF.Run will try to run the async method in line if possible but otherwise, run in BG while blocking current thread in a way that if it is called from UI thread, don't block other JTF from running?

davkean replied:
JTF.Run will not switch to the UI thread without an explicit switch. However, if we're on a background thread and if GetDynamicFileInfoAsync or something further down does switch that's where we run into trouble.

heejaechang replied:
@davkean but that's a problem of using async in a synchronous path and not directly related to JTF. correctly used JTF actually help us from getting into dead lock in those cases.

anyway, this is what @jasonmalinowski asked me to do until methods in IWorkspaceProjectContext is changed to Async and whole call path become async path.

davkean replied:
Yes exactly, sorry - I just had a mental model of how this was implemented under the covers and it didn't match that. We can ignore this whole thread until we flow through async. At that point, I'll call from background thread.

AArnott replied:
@rynowak @heejaechang  `JTF.Run` does *not* transition to the main thread. It invokes the delegate on the caller's thread. The delegate passed to it *may* switch to the main thread.

rynowak replied:
@AArnott - thanks for clearing it up. We were seeing some really wierd behavior whenever I pushed an update through this API. We debugged it an found the real cuprit 👍

heejaechang committed:
fixed the bugs.

heejaechang committed:
add interface and basical implemenetation on some IDocumentServices we discussed yesterday.

still work in progress.

but it should at least now have code to filter out document that doesn't support diagnostics and ignore changes for document that doesn't support changes.

using primary buffer on presentation, I am still working on.

jasonmalinowski commented:
@heejaechang My primary PR has been merged, so you'll want to recreate this PR against dev16.0.x in the Roslyn repo itself.

heejaechang committed:
moved FAR to work on top of new set of APIs.

it should now show things on html correctly

heejaechang committed:
made code path to handle case where mapper returns nothing

heejaechang committed:
added support for code lens.

but found 1 issue. without changing VS side, I can't fix filename. and without fixing filename, navigation won't work properly

heejaechang committed:
trim whitespace before inline text

jasonmalinowski commented:
Closing now in favor of dotnet/roslyn#30733.
